From 5b917ffbddfbb7366d2ad7d5f0e9c016ced29ca5 Mon Sep 17 00:00:00 2001
From: Lukas Czerner <lczerner@redhat.com>
Date: Wed, 15 Jan 2014 17:52:28 +0100
Subject: [PATCH 07/10] misc: Fix get_device_size to work with partitions
 correctly

Currently get_device_size() searches for the size in sysfs specifically
(/sys/block/<devname>/size) however that is not the right palce for
partitions since partitions are subdirectories in /sys/block/<devname>
directory.

So change get_device_size() to get major and minor numbers of the device
in question and use sysfs file /sys/dev/block/<major>:<minor>/size to
get the size.

Signed-off-by: Lukas Czerner <lczerner@redhat.com>
---
 ssmlib/misc.py | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/ssmlib/misc.py b/ssmlib/misc.py
index bfaa1c4..dbfa6b1 100644
--- a/ssmlib/misc.py
+++ b/ssmlib/misc.py
@@ -608,8 +608,9 @@ def is_bdevice(path):
 
 
 def get_device_size(device):
-    devname = device.split('/')[-1]
-    with open("/sys/block/{0}/size".format(devname), 'r') as f:
+    info = os.stat(device)
+    major, minor = divmod(info.st_rdev, 256)
+    with open("/sys/dev/block/{0}:{1}/size".format(major, minor), 'r') as f:
         for line in f:
             size = int(line)/2
             return size
-- 
1.8.3.1

